### ISSUE TITLE
support filtering on relation fields

### ISSUE THREAD

**Issue by jamesdixon** [scoutforpets/bookshelf-jsonapi-params#16]
Currently, the plugin only supports filtering on the top level model. For example, for a `person` that has many `pets`, you can only sort on fields that belong to `person`. If you wanted to find only people that have dogs, you cannot do something like `GET /people?filter[pet.species]=dog`.

**Comment by alechirsch** [comment]
I am currently looking into solving this issue.

**Comment by jamesdixon** [comment]
@alechirsch I did have someone ask in the past to be able to support filtering on relationships n layers deep. I think that's probably a bit overkill and increases the complexity quite a bit. That said, I'd recommend only support relations a single layer deep. Thoughts?

**Comment by alechirsch** [comment]
I have an idea that would allow n layers deep. The implementation that I am considering would allow for both filtering and sorting by related fields n layers deep. The basic idea is to look through all filter and sort fields, split each by "." and maintain a hash of tables required for the filters and sorting. Then loop through the hash and chain "left outer join".

I tested this with raw SQL for single layer only as you can see below

```
select campaigns.*
from campaigns left outer join lookups on campaigns.status_id = lookups.id 
    left outer join clients on campaigns.client_id = clients.id 
    left outer join campaigns_to_forms on campaigns.id = campaigns_to_forms.id
    left outer join forms on campaigns_to_forms.form_id = forms.id
where forms.id = 9
order by clients.name
```

**Comment by jamesdixon** [comment]
@alechirsch sounds good to me! If you can make it work without too much complexity, then go for it!

I actually believe I had started on this a while back, but I recall it being late and producing fairly shitty results, so I stopped :) However, I did come up with a slightly more robust (I think) way to parse the sort parameter. Take a look and see if it's something that should be integrated as you're working through this: https://github.com/scoutforpets/bookshelf-jsonapi-params/commit/fbc75b57c2eb8418433987898222ae5a050ee7bb#diff-1fdf421c05c1140f6d71444ea2b27638R226

Thanks!

**Comment by alechirsch** [comment]
@jamesdixon I am currently developing this feature, as well as #15, on [this branch](https://github.com/alechirsch/bookshelf-jsonapi-params/tree/related-fields-sort-filter). Feel free to look at it, there is still work to be done. I currently have all of the required relationships mapped out and still need to add those relationships to the main query.

The idea is that we would not need to modify any of the other build functions.

**Comment by alechirsch** [comment]
I ran into a problem with anything more than one layer deep. Do you know of a way to access the model object of a relation? I think this is order to look up the relationship data,  which has the tables and foreign key information

**Comment by alechirsch** [comment]
@jamesdixon I figured out how to access the relation models and I now have n-level deep filtering and sorting for relationships. Each relationship type has a special case, currently I have the following relationship types implemented: hasOne, hasMany, belongsTo, belongsToMany. I still need to implement for the three Morph types.

I have the latest code on this [branch](https://github.com/alechirsch/bookshelf-jsonapi-params/tree/related-fields-sort-filter)

**Comment by alechirsch** [comment]
@jamesdixon Did you ever find and answer for your question that you posted [here](http://stackoverflow.com/questions/35997706/reverse-relationship-lookup-with-bookshelf-js-polymorphic-relation?rq=1)?

**Comment by jamesdixon** [comment]
@alechirsch I didn't. See here: https://github.com/tgriesser/bookshelf/issues/1159

**Comment by jamesdixon** [comment]
Closed via #22
